src/dbnode/storage/index.go

```diff
@@ -1074,18 +1073,16 @@ func (i *nsIndex) WarmFlush(
 				zap.Error(err),
 				zap.Time("blockStart", block.StartTime().ToTime()),
 			)
-			continue
 		}
 
-		for _, s := range shards {
-			flushes[shardFlushKey{
-				shardID:    s.ID(),
-				blockStart: block.StartTime(),
-			}] = s
+		for _, t := range i.blockStartsFromIndexBlockStart(block.StartTime()) {
+			for _, s := range shards {
+				s.MarkWarmIndexFlushStateSuccessOrError(t, err)
```

Comment:
So, IIUC, `blockStartsFromIndexBlockStart` returns data block starts between index block start and index block start + index block size. This is important because index block size >= data block size. If that's true, why do we need to `MarkWarmIndexFlushStateSuccessOrError` for each data block start time?

Reply:
That's correct. And the reason is because `MarkWarmIndexFlushStateSuccessOrError` is still marking data blockStarts (not index blockStarts), but each block now just have a flag for both data and index. This is because the state we are tracking is always still at the data block size not index block size. Eg we have 1h block and 2h indexBlock. In-mem we now have:
```
1pm: {dataFlushed: bool, indexFlushed: bool}
2pm: {dataFlushed: bool, indexFlushed: bool}
3pm: {dataFlushed: bool, indexFlushed: bool}
4pm: {dataFlushed: bool, indexFlushed: bool}
...
```
where determining if a given block is eligible for GC we check both `dataFlushed && indexFlushed`. So when an index flush occurs, that actually in this case covers 2 blocks.

This is a somewhat confusing nuance though so if you have any suggestions are making this easier to understand let me know.

Reply:
Yeah, may be worth adding a comment here just clarify